File: Python/PySparkExamples/src/examples/ReadCSV.py
from pyspark.sql import SparkSession
from pyspark import StorageLevel, Row
from pyspark.sql.functions import col, lit, array_contains, split
from pyspark.sql.types import StructType, StructField, StringType, IntegerType, ArrayType, DoubleType, BooleanType

# Creation of spark object
spark = SparkSession.builder \
    .master("local[1]") \
    .appName("PySpark Examples") \
    .getOrCreate()

file_path = "../../config/weather.csv"
df = spark.read.csv(file_path)
df.printSchema()

df = spark.read.format("csv").load(file_path)
#or
#df = spark.read.format("org.apache.spark.sql.csv").load(file_path)
df.printSchema()

#Using Header record for column names
df2 = spark.read.option("header",True).csv(file_path)

#Read multiple CSV files
#df = spark.read.csv("file_path,file_path,file_path")

#Read all CSV files in a directory
#df = spark.read.csv("Folder path")

#Options while reading CSV file
df3 = spark.read.options(delimiter=',') \
    .csv(file_path)

#inferSchema
df4 = spark.read.options(inferSchema='True',delimiter=',') \
    .csv(file_path)

#or
df4 = spark.read.option("inferSchema",True) \
    .option("delimiter",",") \
    .csv(file_path)

#header
df3 = spark.read.options(header='True', inferSchema='True', delimiter=',') \
    .csv(file_path)


#Reading CSV files with a user-specified custom schema
schema = StructType() \
    .add("RecordNumber",IntegerType(),True) \
    .add("Zipcode",IntegerType(),True) \
    .add("ZipCodeType",StringType(),True) \
    .add("City",StringType(),True) \
    .add("State",StringType(),True) \
    .add("LocationType",StringType(),True) \
    .add("Lat",DoubleType(),True) \
    .add("Long",DoubleType(),True) \
    .add("Xaxis",IntegerType(),True) \
    .add("Yaxis",DoubleType(),True) \
    .add("Zaxis",DoubleType(),True) \
    .add("WorldRegion",StringType(),True) \
    .add("Country",StringType(),True) \
    .add("LocationText",StringType(),True) \
    .add("Location",StringType(),True) \
    .add("Decommisioned",BooleanType(),True) \
    .add("TaxReturnsFiled",StringType(),True) \
    .add("EstimatedPopulation",IntegerType(),True) \
    .add("TotalWages",IntegerType(),True) \
    .add("Notes",StringType(),True)

df_with_schema = spark.read.format("csv") \
    .option("header", True) \
    .schema(schema) \
    .load("../../config/zipcodes.csv")


#Write PySpark DataFrame to CSV file
df.write.option("header",True) \
    .csv("../../config/zipcodes_1.csv")

#Options
df2.write.options(header='True', delimiter=',') \
    .mode('overwrite')\
    .csv("../../config/zipcodes_1.csv")

#Saving modes
df2.write.mode('overwrite').csv("../../config/zipcodes_1.csv")
#you can also use this
df2.write.format("csv").mode('overwrite').save("../../config/zipcodes_1.csv")

File: Scala/Spark-Scala-Examples/src/main/scala/com/examples/ReadCSV.scala
package com.examples
import org.apache.spark.rdd.RDD
import org.apache.spark.sql.{SaveMode, SparkSession}
import org.apache.spark.sql.types.{BooleanType, DoubleType, IntegerType, StringType, StructType}

object ReadCSV extends App {

  val spark:SparkSession = SparkSession
    .builder()
    .master("local[1]")
    .appName("Spark-Scala-Examples")
    .getOrCreate()


  //Load CSV file into RDD
  val rddFromFile = spark.sparkContext.textFile("./config/weather.csv")

  val rdd = rddFromFile.map(
    f =>
      {
        println(f(0) + " " + f(1))
      }
  )

  rdd.foreach(
    f =>
      {
        println(f)
      }
  )

  val df1 = spark.read.csv("./config/zipcodes.csv")
  df1.printSchema()

  val d2 = spark.read.option("header",true)
    .csv("./config/zipcodes.csv")

  val df3 = spark.read.options(Map("delimiter"->","))
    .csv("./config/zipcodes.csv")

  val df4 = spark.read.options(Map("inferSchema"->"true","delimiter"->",","header"->"true"))
    .csv("./config/zipcodes.csv")

  val schema = new StructType()
    .add("RecordNumber",IntegerType,true)
    .add("Zipcode",IntegerType,true)
    .add("ZipCodeType",StringType,true)
    .add("City",StringType,true)
    .add("State",StringType,true)
    .add("LocationType",StringType,true)
    .add("Lat",DoubleType,true)
    .add("Long",DoubleType,true)
    .add("Xaxis",IntegerType,true)
    .add("Yaxis",DoubleType,true)
    .add("Zaxis",DoubleType,true)
    .add("WorldRegion",StringType,true)
    .add("Country",StringType,true)
    .add("LocationText",StringType,true)
    .add("Location",StringType,true)
    .add("Decommisioned",BooleanType,true)
    .add("TaxReturnsFiled",StringType,true)
    .add("EstimatedPopulation",IntegerType,true)
    .add("TotalWages",IntegerType,true)
    .add("Notes",StringType,true)
  val df_with_schema = spark.read.format("csv")
    .option("header", "true")
    .schema(schema)
    .load("./config/zipcodes.csv")
  df_with_schema.printSchema()
  df_with_schema.show(false)


  df1.write.option("header","true")
    .csv("./config/zipcodes_output")

  df1.write.option("header",true)
    .csv("./config/zipcodes_output1")

  df1.write.mode(SaveMode.Overwrite).csv("./config/zipcodes_output")


}
